src/EditorFeatures/CSharpTest/CodeActions/ConvertLocalFunctionToMethod/ConvertLocalFunctionToMethodTests.cs

```diff
@@ -505,10 +505,11 @@ public async Task TestCaretPositon()
             await TestAsync("C Local[||]Function(C c)");
             await TestAsync("C [|LocalFunction|](C c)");
             await TestAsync("C LocalFunction[||](C c)");
-            await TestMissingAsync("C Local[|Function|](C c)");
+            await TestAsync("C Local[|Function|](C c)");
```

Comment:
was thi change intentional?  seems somewhat undesirable.  i think selections taht cut off words are unlikely to be intended to be used for a refactoring like this.

Reply:
It was intentional. If we want to enable situations similar to `"C L[|ocalFunction|](C c)"` without rather complicated logic we need to allow that one as well. Also, it's been discussed somewhere above with @mavasani .

Reply:
> If we want to enable situations similar to "C L[|ocalFunction|](C c)" without rather complicated logic 

hrmm... i'm not really getting the complexity.  I think in the helper it would be pretty simple.  specifically, if the selection is non-empty, it should not start/end in the middle of a token's span.

Reply:
Sure, but then we loose the refactoring in situations such as `"C L[|ocalFunction|](C c)"` or generally (since we're creating helper) everytime the user misses one/few chars in the beginning/end. Which I felt was undesirable. 

There was actually a version that checked the exact spans but got feedback to remove it.

Reply:
> Sure, but then we loose the refactoring in situations such as "C L[|ocalFunction|](C c)" or generally (since we're creating helper) everytime the user misses one/few chars in the beginning/end. Which I felt was undesirable.

That doesn't seem like a problem to me.  Common patterns are likely to be overselection, or double-clicking on the id.  having a partial selection but still expecting it to work seems like a stretch.  Or, in other words, i would have no problem explaining to a user: it didn't show up because you selected some random substring, and we figure it was highly unlikely ou wanted this.

> There was actually a version that checked the exact spans but got feedback to remove it.

Yes.  I would not check exact spans.  I'd have a few overall behaviors that seemed reasonable.  i.e.:

1. was the whole node selected.  
2. was the caret somewhere on the 'header' of the item (where header is defined on a per type basis**).
3. if there's a selection, did it select the header or name of the item.

These all seem to match a reasonable expectation of what people may want to do.

--

** We already have the 'header' concept for things like 'type declarations' where we understand that it may be somewhere in the modifiers, or the name, etc. etc.  See `IsOnTypeHeader` as an example.

Reply:
I like Petr’s current implementation, it’s simpler and avoids users having to ensure that a precise span was selected. This was one of core feedback that missing a part of selection like say a closing brace suddenly causes refactoring to not show up. In current implementation, selected span is mapped to the most logical piece of code associated with it, and customers will not complain if the refactoring does show up here. This gets even more trickier for customers for refactorings where it is not easily evident what the precise span of a node would be get the refactoring, say for refactorings within expressions. If this approach turns out to generate noise, we can make it stricter as you suggest, but otherwise erring on side of being looser seems preferable.

Reply:
> and avoids users having to ensure that a precise span was selected

Note: i'm in favor of not requiring the user to have a precise span.  however, there's a range between: 

1. the span must be precisely this things, and
2. the span can be anything

My only gripe here is allowing partial token selection.  It's *that* piece in particular that seems very suspect.  I'm otherwise in favor of the loosening here.  I'm just trying to strike a balance between "we were too strict before" and "we will allow anything".

> This gets even more trickier for customers for refactorings where it is not easily evident what the precise span of a node would be get the refactoring, say for refactorings within expressions.

I'm not certain what this has to do with other refactorings though.  I'm specifically talking about this *this* particular refactoring, and how i don't think it's sensible that a partial identifier selection should trigger it.

Reply:
> If this approach turns out to generate noise, we can make it stricter as you suggest,

I honestly think no customer would be surprised by the behavior that if they're selecting things that the selection at least lands *around* tokens and not *in the middle of individual tokens*.    Consider the tokens that can even be multiple characters.  They are:

1. keywords
2. identifiers
3. literals
4. multi-char ops.

Would a user really expect refactoring if selecting `pu[|blic stat|]ic`   or selecting `void LocalFun[ltion(int par|]am`?

The same would apply for selecting a portion of a literal, or having the selection land like this  `[|++=|]=`.  

In these cases, i think you'd only be cutting *into* tokens for normal textual work.  

Note: this only applies to *selections*.  for things like caret placement, i think we should be permissive.  But for selections, we should assume bisected tokens are *not* an intent to refactor.

Reply:
> That doesn't seem like a problem to me. Common patterns are likely to be overselection, or double-clicking on the id. having a partial selection but still expecting it to work seems like a stretch. 

I see that but what's the downside to letting it show up? If we agree that under-selection shouldn't be accepted then we shouldn't use it anywhere (sans for things like renames) for the sake of consistency. And showing more refactorings is IMO better than showing no/very little.

Later when we start working on how to order the refactorings we can de-prioritize the ones that were undermatched but I don't think that excluding them alltoghether is beneficia. 

> Yes. I would not check exact spans. I'd have a few overall behaviors that seemed reasonable. i.e.:

The mentioned exact match did exactly what you described. Either the whole node had to be selected, or carrot somewhere in the header (function's name in this case) or the whole header selected (again, function's name) (it might also be reasonable to allow if the whole method's declaration (type + name + parames is selected, but that's details).

Reply:
> I see that but what's the downside to letting it show up? If we agree that under-selection shouldn't be accepted then we shouldn't use it anywhere (sans for things like renames) for the sake of consistency.

First, i'm amenable to this argument.  I'd personally say that the downside of allowing partial-token-selectoin is that i think it will bite us later on as we move more and more refactorings over to these principles.  For LocalFunction it's not that big an issue.  But for things like 'expressions' i think it can and will lead to annoying scenarios with too much showing.

>  And showing more refactorings is IMO better than showing no/very little.

I actually highly disagree with this.  Both are problematic, and neither is necessarily better.  We actually spent a huge amount of time dealing with teh reverse problem (i.e. too many refactoring showing, or refactoring showing up innapropriately).   It was actually something we got a lot of feedback on in the past.

I do think we may have swung the needle too far in the other direction, so i'm not at all opposed to making the changes you're making.  I'm just trying to find the appropriate 'sweet spot' that handles the cases that aare high value, using a consistent set of principles.  

> The mentioned exact match did exactly what you described.

Ah, i imisunderstood what you meant by 'exact match'.  Thanks for the clarification.

--

Like i said above, i thnk what you have is almost ideal for me.  I'd just tweak one thing slightly. :)

Reply:
Let me summarize here: all of us are have the same stance on majority of the scenarios here, including when there is a selection that spans a node partially, but none of the tokens are partially selected. We have a slight disagreement on when there is a selection that partially spans the first and/or last token, but IMO that scenario is much less critical then partial node selection where there are no partially selected tokens. I don't think the customers will complain either way. I think we should just take the simplest approach for this. I have a soft preference towards erring on the side of showing the refactorings for the parent node and am fine to switch to the other side if there is feedback that this is noisy. @heejaechang @kendrahavens @mikadumont any opinions here?

Reply:
@mavasani I'm amenable to that.  As long as it's a conscientious choice, and didn't fall out accidentally :)  

Personally, i would err on *not* showing here just because i think it will be desirable *later* on.  For example, when people select portions of the inside of strings, i don't necessarily want them prompted with a lightbulb that contains strange entries.  However, i'm not going to die on this hill.  If you guys talk about it and decide to go this way, that's ok with me.  

Let's just try to keep this case in our minds going forward so we don't accidentally have any behavior in the future we don't like.

Thanks!